device_backends/xdma/src/XdmaBackend.cc

```diff
@@ -0,0 +1,126 @@
+#include "XdmaBackend.h"
+
+#include <boost/make_shared.hpp>
+#include <iomanip>
+
+namespace ChimeraTK {
+
+  XdmaBackend::XdmaBackend(std::string devicePath, std::string mapFileName)
+  : NumericAddressedBackend(mapFileName), _devicePath(devicePath) {}
+
+  XdmaBackend::~XdmaBackend() {}
+
+  void XdmaBackend::open() {
+#ifdef _DEBUG
+    std::cout << "open xdma dev: " << _devicePath << std::endl;
+#endif
+    if(_ctrlIntf) {
+      if(isFunctional()) {
+        return;
+      }
+      close();
+    }
+    // TODO: retrieve mmap_(offs,size) from map file
+    _ctrlIntf.emplace(_devicePath);
+
+    // Build vector of max. 4 DMA channels
+    _dmaChannels.clear();
+    for(size_t i = 0; i < 4; i++) {
+      try {
+        _dmaChannels.emplace_back(_devicePath, i);
+      }
+      catch(const runtime_error&) {
+        break;
+      }
+    }
+    _hasActiveException = false;
+  }
+
+  void XdmaBackend::close() {
+    _ctrlIntf.reset();
+    _dmaChannels.clear();
+  }
+
+  bool XdmaBackend::isOpen() { return !!_ctrlIntf; }
+
+  bool XdmaBackend::isFunctional() const {
+    if(!_ctrlIntf || _hasActiveException) {
+      return false;
+    }
+    // TODO implement me
+    return true;
+  }
+
+  XdmaIntfAbstract* XdmaBackend::_intfFromBar(uint64_t bar) {
+    if(bar == 0) {
+      return _ctrlIntf ? dynamic_cast<XdmaIntfAbstract*>(&_ctrlIntf.value()) : nullptr;
+    }
+    // 13 is magic value for DMA channel (by convention)
+    // We provide N DMA channels starting from there
+    const ssize_t dmaChIdx = static_cast<ssize_t>(bar) - 13;
+    if(dmaChIdx < 0 || dmaChIdx >= static_cast<ssize_t>(_dmaChannels.size())) {
+      auto exc_string =
+          std::string{"Couldn't find XDMA channel for BAR value "} + std::to_string(static_cast<int>(bar));
```

Comment:
Opps, this static cast was overlooked in the merge with master. When bar was still uint8_t it would have been treated as a character, so the cast was necessary. With uin64_t this is not needed any more. My bad.